fhirpipe/transform/fhir.py

```diff
@@ -1,122 +1,199 @@
-                col_name = new_col_name(inp["script"], col_name)
-            cols_to_fetch[0].append(col_name)
+    # Remove duplicates in fhir object
+    fhir_object = clean_fhir_object(fhir_object)
 
-        # Else retrieve the static value
-        else:
-            cols_to_fetch[1].append(inp["staticValue"])
+    return fhir_object
 
-    if structure["mergingScript"]:
-        result = row[new_col_name(structure["mergingScript"], cols_to_fetch)]
-    else:
-        result = [row[c] for c in cols_to_fetch[0]]
-        result.extend(cols_to_fetch[1])
 
-    # Unlist
-    if isinstance(result, list) and len(result) == 1:
-        result = result[0]
+def build_fhir_object(row, path_attributes_map, index=None):
+    """ Function that actually builds a nested object from the dataframe row and the mapping.
+    Note that it can be used to build only a subpart of a fhir instance.
+    """
+    fhir_object = recursive_defaultdict()
+    arrays_done = set()
 
-    if isinstance(result, tuple):
-        for ind in indices:
-            result = result[ind]
+    for path, attr in path_attributes_map.items():
+        if attr["inputs"]:
 
-    if isinstance(result, tuple):
-        if not all([el == result[0] for el in result[1:]]):
-            raise Exception(f"Cannot have as input a list with different values ({result}).")
-        result = result[0]
+            split_path = path.split(".")
+            position_ind = get_position_first_index(split_path)
 
-    return result
+            if position_ind is None:
+                val = fetch_values_from_dataframe(row, attr["inputs"], attr["mergingScript"])
+                if isinstance(val, tuple) and len(val) == 1:
+                    insert_in_fhir_object(fhir_object, path, val[0])
+                elif isinstance(val, tuple) and index is not None:
+                    insert_in_fhir_object(fhir_object, path, val[index])
+                else:
+                    insert_in_fhir_object(fhir_object, path, val)
+
+            else:
+                # Handle array case
+                array_path = ".".join(split_path[: position_ind + 1])
+                # If this path was already processed before, skip it
+                if array_path in arrays_done:
+                    continue
+                # We retrieve all the attributes that start with the array path (with index)
+                attributes_in_array = {
+                    remove_root_path(path, position_ind + 1): attr
+                    for path, attr in path_attributes_map.items()
+                    if path.startswith(array_path)
+                }
+                # Build the array of sub fhir object
+                array = handle_array_attributes(attributes_in_array, row)
+                # Insert them a the right position
+                if array:
+                    insert_in_fhir_object(fhir_object, ".".join(split_path[: position_ind]), array)
+                arrays_done.add(array_path)
```

Comment:
Maybe you'd like to delete all the keys that you've processed in `path_attributes_map` instead of having an extra variable `arrays_done` to record what you've done?

Reply:
I'd like to but it's dangerous to remove elements in a dictionary while looping on it (so dangerous that Python forbids it, I think)